README.md
# myterm

### _(command line api for playing with iTerm)_

## About

`myterm` is a command line interface for customizing iTerm using AppleScript.  Its most useful feature is that it can throw up a big message in the background of your terminal tab to tell you clearly which tab it is.




## Usage

`myterm` attempts to be as self-documenting as possible.  Try `myterm -h` for help.

The most common use case for me personally is usually something like: 

    myterm bg ohai

where "ohai" is some label that I want to appear in the background of the iTerm window.

More ambitiously, I might do:

    myterm bg --exec node server.js

which will display the string "node server.js" up in the background and then run that.




## Requirements

`myterm` needs ImageMagick and its `convert` executable to be in your path.

Installing ImageMagick on your mac is "easy" with macports, provided that you don't run into hiccups.  (But be prepared for it to take something like5 minutes.)

`myterm` needs some font file to use.  It has an installer that will try to grab the [Simple Life](http://www.dafont.com/simple-life.font) font by Michael Strobel.

Additionally it needs whatever its gem dependencie(s) are (which at the time of this writing are: rb-appscript, highline, rmagick).  Installing the myterm ruby gem should install these dependencies.  If you are developing myterm you could also install them with "bundle install".





## Installation

`myterm` is a rubygem.  If you are installing it from a git checkout
(which at the time of this writing is the only way to install it):

    mkdir ~/src; cd ~/src
    git clone git@github.com:hipe/myterm.git; cd myterm
    gem build myterm.gemspec
    gem install myterm-VERSION.gem # where VERSION is whatever version was built.



## Support

I want this to work for you.  Please contact me via email if it does not!




## Credits

`myterm` started life as an adaptation of Dmytro Shteflyuk's script that he used at Scribd as described  [here](http://kpumuk.info/mac-os-x/how-to-show-ssh-host-name-on-the-iterms-background/).  Thank you Dmytro!


CHANGELOG.md


## for version 0.0.4 -

### 2012-01-20

  * fix issue with tmux sessions
  * overhaul whitespace & naming conventions



._repath.rb
#!/usr/bin/env ruby

payload = lambda do

  dir = File.expand_path('../bin', __FILE__)
  File.directory?(dir) or return {
    :message => "not a directory, won't add to PATH: #{dir}"
  }

  path = ENV['PATH']
  parts = path.split(':')
  case parts.index(dir)
  when nil
    {
      :new_path => ([dir] + parts).join(':'),
      :message  => "prepending bin folder to the beginning of the PATH.",
      :success  => true
    }
  when 0
    { :message  => "already at front of path: \"#{dir}\"",
      :success  => true
    }
  else
    {
      :new_path => ([dir] + parts.reject{ |x| x == path }).join(':'),
      :message  => "rewriting path to have bin folder at the beginning",
      :success  => true
    }
  end

end.call

payload[:message] and $stdout.puts "echo #{payload[:message].inspect}"
payload[:new_path] and $stdout.puts "export PATH=\"#{payload[:new_path]}\""
$stdout.puts(payload[:success] ? "echo 'hack done.'" : "echo 'hack failed.'")


lib-iterm.sh
#!/usr/bin/env bash

## NOTE for `iterm` users -- this file is not used: it is here for reference.
## (it lives in a different git repository)

# adapted from http://kpumuk.info/mac-os-x/how-to-show-ssh-host-name-on-the-iterms-background/

HEIGHT_TALL=750
WIDTH_NARROW=600
WIDTH_SCROLLBAR=15
HEIGHT_TITLEBAR_TABS=44

# arr=($(iterm_bounds_get))
function iterm_bounds_get {
  local size=( $(
    osascript -e "
      tell application \"iTerm\"
        get the bounds of the first window
      end tell" | tr ',' ' '
  ) )
  echo "${size[@]}"
}

function iterm_bounds_set {
  if [[ ! "$1,$2,$3,$4" =~ "^[0-9]+,[0-9]+,[0-9]+,[0-9]+$" ]] ; then
    echo "bad args to item_bounds_set: ($1, $2, $3, $4)"
    return
  fi
  local exec_me="
    tell application \"iTerm\"
      set the bounds of the first window to {$1, $2, $3, $4}
    end tell
  "
  osascript -e "$exec_me"
}

function iterm_dimensions_get {
  local size=($(iterm_bounds_get))
  local x1=${size[0]} y1=${size[1]} x2=${size[2]} y2=${size[3]}
  local w=$(( $x2 - $x1 - $WIDTH_SCROLLBAR))
  local h=$(( $y2 - $y1 - $HEIGHT_TITLEBAR_TABS))
  echo "${w}x${h}"
}

function iterm_dimensions_set {
  local w=$1 h=$2
  local b=($(iterm_bounds_get))
  local x=${b[0]} y=${b[1]}
  iterm_bounds_set $x $y $w $h
}

# check to see if we have the correct terminal for doing this kind of thing
# this won't work when we sudo something because TERM_PROGRAM isn't picked up
function iterm_ok {
  if [ "$(tty)" == 'not a tty' ] || [ "$TERM_PROGRAM" != "iTerm.app" ] ; then
    echo ''
  else
    echo 'ok'
  fi
}

function iterm_set {
  local prop=$1
  local val=$2
  local tty=$(tty)
  osascript -e "
    tell application \"iTerm\"
      repeat with theTerminal in terminals
        tell theTerminal
          try
            tell session id \"$tty\"
              set bounds of window 1 to \"$val\"
            end tell
            on error errmesg number errn
          end try
        end tell
      end repeat
    end tell
  "
}

function iterm_bg_image_make {
  local text1=$1
  local text2=$2
  local color_bg=${3:-#000000}
  local color_fg=${4:-#662020}
  local dimensions=$(iterm_dimensions_get)
  local font_ttf=${5:-$HOME/.bash/resources/SimpleLife.ttf}
  local font_points=${6:-60}
  local font_style=${7:-Normal} # Font style (Any, Italic, Normal, Oblique)
  local gravity=${8:-NorthEast}
      # Text gravity (NorthWest, North, NorthEast,
      # West, Center, East, SouthWest, South, SouthEast)
  local offset1=${8:-20,10}
  local offset2=${9:-20,80}
  local outpath="/tmp/iTermBG.$$.png"
  convert \
    -size "$dimensions" xc:"$color_bg" -gravity "$gravity" -fill "$color_fg" \
    -font "$font_ttf" -style "$font_style" -pointsize "$font_points" \
    -antialias -draw "text $offset1 '$text1'" \
    -pointsize 30 -draw "text $offset2 '$text2'" \
    "$outpath"
  echo $outpath
}

function iterm_bg_image_delete {
  local path=${1:-/tmp/itermBG.$$.png}
  rm $path
}

function iterm_bg_image_empty {
  local opath="/tmp/iTermBG.empty.png"
  if [ ! -f /tmp/iTermBG.empty.png ]; then
    local dims=$(iterm_dimensions_get)
    convert -size "$dims" xc:"#000000" "$opath"
  fi
  echo $opath
}

function iterm_bg_image_set {
  local tty=$(tty)
  osascript -e "
    tell application \"iTerm\"
      repeat with theTerminal in terminals
        tell theTerminal
          try
            tell session id \"$tty\"
              set background image path to \"$1\"
            end tell
            on error errmesg number errn
          end try
        end tell
      end repeat
    end tell
  "
}

function iterm_bg_color_set {
  local tty=$(tty)
  osascript -e "
    tell application \"iTerm\"
      repeat with theTerminal in terminals
        tell theTerminal
          try
            tell session id \"$tty\"
              set background image path to \"\"
              set background color to \"$1\"
            end tell
            on error errmesg number errn
          end try
        end tell
      end repeat
    end tell
  "
}

function iterm_window_title_set {
  osascript -e "
    tell application \"iTerm\"
      set the name of the first window to \"$1\"
    end tell
  "
}


lib/myterm/cli.rb
require File.expand_path('../vendor/skylab/face/cli', __FILE__)
require File.expand_path('../api', __FILE__)
require 'open3'


module Skylab::Myterm
  module PathPrettifier
    HOME_DIR_RE = /\A#{Regexp.escape(ENV['HOME'])}/
    def pretty_path path
      path.sub(HOME_DIR_RE, '~')
    end
  end

  class ::Skylab::Face::Command
    include PathPrettifier # eew
  end

  class Cli < ::Skylab::Face::Cli
    include PathPrettifier

    version { ::Skylab::Myterm.version }

    o(:bounds) do |o|
      syntax "#{invocation_string} [x y width height]"
      o.banner = "gets/sets the bounds of the terminal window\n#{usage_string}"
    end

    def bounds o, *a
      case a.length
      when 4
        begin ; iterm.bounds = a ; rescue ValidationError => e ; return usage e ; end
      when 0
        @err.puts iterm.bounds.inspect
      else
        return usage("bad number of args #{a.length}: expecting 0 or 4")
      end
    end

    o(:'bg') do |o, req|
      syntax "#{invocation_string} [opts] [<text> [<text> [...]]]"
      o.banner = "Generate a background image with certain text for the terminal\n#{usage_string}"
      o.on('-e', '--exec <cmd ...>', 'Execute <cmd ...> in shell, also use it as text for background.') { }
      o.on('-o', '--opacity PERCENT', "Percent by which to make image background opaque",
        "(0%: tranparent. 100%: solid.  Default: solid)") { |amt| req[:alpha_percent] = amt }
      req[:font_file] = DEFAULT_FONT_FILE
      o.on('--font FONTFILE.ttf', "font to use (default: #{pretty_path(req[:font_file])})") do |path|
        req[:font_file] = path
      end
      req[:fill] = '#662020'
      o.on('--fill[=COLOR]', "Write text in this color (default: #{req[:fill].inspect})",
        "(when present but with no value, will use \"Text/Normal\" setting of current iTerm tab)" ) do |v|
        req[:fill] = v || lambda { |img| img.iterm.session.foreground_color.to_hex }
      end
      o.on('-v', '--verbose', 'Be verbose.') { req[:verbose] = true }
    end

    def before_parse_bg req, args
      idx = args.index { |s| %w(-e --exec).include?(s) } or return true
      req[:_exec_this] = args[(idx+1)..-1]
      args.replace idx == 0 ? [] : args[0..(idx-1)]
      true
    end
    protected :before_parse_bg

    def bg req, *args
      if args.empty?
        if req[:_exec_this]
          args.any? and fail("logic error -- see before_parse_bg.")
          args = req[:_exec_this]
        else
          return get_background
        end
      end
      check_font(req) or return
      img = ImageBuilder.build_background_image(iterm, args, req) or return false
      req[:verbose] and @err.puts "(bg_color: #{img.background_color.inspect})"
      outpath = "#{IMG_DIRNAME}/#{IMG_BASENAME}.#{Process.pid}.png"
      img.write(outpath)
      req[:verbose] and @err.puts "(setting background image to: #{outpath})" # doesn't care if --verbose
      iterm.session.background_image_path = outpath
      if req[:_exec_this]
        @err.puts "(#{program_name} executing: #{req[:_exec_this].join(' ')})"
        exec(req[:_exec_this].join(' '))
      end
      true
    end

    DEFAULT_FONT_FILE = "#{ENV['HOME']}/.fonts/MytermDefaultFont.ttf"
    IMG_DIRNAME  = '/tmp'
    IMG_BASENAME = 'iTermBG'

  private

    def check_font req
      File.exist?(req[:font_file]) and return true
      if req[:font_file] == DEFAULT_FONT_FILE
        return maybe_download_font req
      else
        font_not_found req
      end
    end

    DEFAULT_FONT_URL                = 'http://img.dafont.com/dl/?f=simple_life'
    DEFAULT_FONT_FILE_NOT_SIMLINKED = "#{ENV['HOME']}/.fonts/SimpleLife.ttf"

    def maybe_download_font req
      target = DEFAULT_FONT_FILE_NOT_SIMLINKED
      File.exist?(target) and return true
      $stdin.tty? && $stdout.tty? or return font_not_found(req)
      @err.write "Font file #{pretty_path(req[:font_file])} not found.  "
      require 'highline'
      require 'fileutils'
      HighLine.new.agree("Let #{program_name} download it? (Y/n) (recommended: yes)") or return false
      outfile = DEFAULT_FONT_FILE_NOT_SIMLINKED.sub(/\.ttf$/, '.zip')
      font_dir = File.dirname(outfile)
      File.directory?(font_dir) or FileUtils.mkdir_p(font_dir, :verbose => true)
      cmds = ["wget -O #{outfile} #{DEFAULT_FONT_URL}"]
      cmds.push "cd #{font_dir}"
      cmds.push "unzip #{outfile}"
      cmds.push "ln -s #{DEFAULT_FONT_FILE_NOT_SIMLINKED} #{DEFAULT_FONT_FILE}"
      cmds.push("echo 'finished installing for #{program_name}: #{pretty_path(DEFAULT_FONT_FILE_NOT_SIMLINKED)}. " <<
        "Please try using it again.'")
      @err.puts(cmd = cmds.join(' ; '))
      exec(cmd)
    end

    def font_not_found req
      @err.puts "font file not found: #{pretty_path(req[:font_file])}"
      req.command.usage
      return false
    end

    def iterm
      @iterm ||= ItermProxy.new
    end

    def get_background
      @err.puts "tty: #{iterm.session.tty}"
      @err.puts "background_image: #{iterm.session.background_image_path.inspect}"
      @err.puts "background_color: #{iterm.session.background_color}"
    end
  end
end



lib/myterm/api.rb
module Skylab; end

module Skylab::Myterm

  class ValidationError < RuntimeError ; end

  module Color
    def alpha= mixed
      if mixed.kind_of?(String)
        md = /\A(\d+(?:\.\d+)?)%?\z/.match(mixed) or
          raise ValidationError.new("invalid format for percent #{val.inspect} -- expecting e.g. \"58%\"")
        mixed = md[1].to_f
      end
      (0.0..100.0).include?(mixed) or
        raise ValidationError.new("Percent value (#{mixed}%) must be between 0 and 100 inclusive.")
      self[3] = ChannelScalarNormalized[mixed / 100.0]
    end

    def to_hex
      '#' + self.map{ |x| int_to_hex(x) }.join('')
    end

    TARGET_PLACES = 2  # each component of an #rrggbb hexadecimal color has 2 places

    DIVISOR = 16 ** TARGET_PLACES

    def int_to_hex int
      int.respond_to?(:to_hex) and return int.to_hex
      (int.to_f / DIVISOR).round.to_s(16).rjust(TARGET_PLACES, '0')
    end
  end

  class << Color
    def [] obj
      obj.extend self
    end

    def dup color
      self[color.dup]
    end
  end

  module ChannelScalarNormalized
    def to_hex
      (('ff'.to_i(16).to_f * self).to_i).to_s(16).rjust(2, '0')
    end
  end

  class << ChannelScalarNormalized
    def [] obj
      obj.extend self
    end
  end

  class ImageBuilder
    def build_text_drawing img
      @lines.empty? and return fail("foo")
      #@todo setters for everything etc
      draw = Magick::Draw.new
      draw.gravity         = @opts[:gravity] || Magick::NorthEastGravity
      @opts[:fill] ||= '#662020'
      @opts[:fill].kind_of?(Proc) and @opts[:fill] = @opts[:fill].call(self)
      draw.fill            = @opts[:fill]
      draw.font            = @opts[:font] || "#{ENV['HOME']}/.fonts/SimpleLife.ttf"
      draw.font_style      = @opts[:font_style] || Magick::NormalStyle
      draw.pointsize       = @opts[:point_size] || 60
      draw.text_antialias  = @opts.key?(:text_antialias) ? @opts[:text_antialias] : true
      draw.annotate(img, 0,0,20,10, @lines.first)
      if @lines.length > 1
        second_line = @lines[1..-1].join(' ')
        draw.annotate(img, 0,0,20,80, second_line) do
          self.pointsize = 30
        end
      end
      nil # draw not needed at this point
    end
    private :build_text_drawing

    def initialize iterm, lines, opts
      @iterm, @lines, @opts = [iterm, lines, opts]
    end

    attr_reader :iterm

    def run
      require 'RMagick'
      bg_color = Color.dup(@iterm.session.background_color)
      @opts.key?(:alpha_percent) and bg_color.alpha = @opts[:alpha_percent]
      img = Magick::Image.new(500, 300) do # copying over hard-coded dimensions from original Dmytro script
        self.background_color = bg_color.to_hex
      end
      build_text_drawing img
      img
    end
  end

  class << ImageBuilder
    def build_background_image iterm, lines, opts
      new(iterm, lines, opts).run
    end
  end

  class ItermProxy
    # ItermProxy is a wrapper around everything Iterm to the extent that its AppleScript interface supports
    #
    MIN_LEN = 50

    def app
      @app ||= begin
        require 'appscript'
        Appscript.app('iTerm')
      end
    end
    private :app

    def invalid msg
      raise ValidationError.new(msg)
    end
    private :invalid

    def bounds= arr
      x = arr.detect { |i| i.to_s !~ /^\d+$/ } and return invalid("expecting digit had #{x.inspect}")
      x = arr[2,2].detect { |i| i.to_i < MIN_LEN } and return invalid("too small: #{x} (min: #{MIN_LEN})")
      app.windows[0].bounds.set arr
    end

    def bounds
      app.windows[0].bounds.get
    end

    def session
      @session ||= begin
        tty = `tty`.strip
        t = s = 0
        catch(:break) do
          (1..app.terminals.count).each do |i|
            (1..app.terminals[i].sessions.count).each do |j|
              if tty == app.terminals[i].sessions[j].tty.get
                throw(:break, SessionProxy.new(app, i, j))
              end
            end
          end
          nil
        end or fail("couldn't ascertain current session!")
      end
    end
  end

  module AppscriptDelegator
    def delegated_attr_readers *list
      list.each do |property|
        lambda do |_property|
          define_method(_property) do
            resource.send(_property).get
          end
        end.call(property)
      end
    end

    def delegated_attr_writers *list
      list.each do |property|
        lambda do |_property|
          define_method("#{_property}=") do |val|
            resource.send(_property).set val
          end
        end.call(property)
      end
    end

    def delegated_attr_accessors *list
      delegated_attr_readers(*list)
      delegated_attr_writers(*list)
    end
  end

  class ItermProxy::SessionProxy
    extend AppscriptDelegator

    delegated_attr_accessors :background_image_path

    def background_color
      Color[resource.background_color.get]
    end

    def foreground_color
      Color[resource.foreground_color.get]
    end

    def initialize app, term_idx, session_idx
      @app, @term_idx, @sessions_idx = [app, term_idx, session_idx]
    end

    def resource
      @app.terminals[@term_idx].sessions[@sessions_idx]
    end

    delegated_attr_readers :tty
  end
end

module Skylab
  class << Myterm
    def version
      File.read(File.expand_path('../../../VERSION', __FILE__))
    end
  end
end

